core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -589,4 +603,20 @@ private static PartitionSpec freshSpec(int specId, Schema schema, PartitionSpec
     }
     return builder.build();
   }
+
+  private static String publishedWapId(Snapshot snapshot) {
```

Comment:
I don't think that this change needs to alter `TableMetadata`. All it does is create a new snapshot with the changes from another snapshot and set the current snapshot. All that is done by `MergingSnapshotProducer` already.

Reply:
these two methods are creating an index to lookup snapshots by published wap.ids ..to protect from duplicates. Similar to what `indexSnapshots()` does. It only reads existing snapshot summaries. The metadata creation itself is fully delegated to `MergingSnapshotProducer` .. do you think this can be achieved differently?

Reply:
Yeah, I think these should be in a utility class for WAP pattern to avoid having this class do too much.

Reply:
will do.

Reply:
done

Reply:
It doesn't look like these changes made it into the PR yet. I still see these methods in `TableMetadata`.

Reply:
I commented on this elsewhere in SnapshotManager .. mentioning it here.. 

 I'v moved `isWapIdPublished()` to a `WapUtil` class (which is now used by SnapshotManager.apply ). Although, we want to keep the complexity of the check to O(1), so a `snapshotsByWapId` (or equivalent) index would need to be built once for each table state. Additionally, Util classes shouldn't be keeping state, as a consequence i'v exposed snapshotsByWapId() as is in TableMetadata which is used by WapUtil from within the convenience method: isWapIdPublished. So the state is maintained by TableMeta but the explicit check is performed by WapUtil.. Lemme know what you think. These two private methods now that they are only used to build the index in TableMeta.

Reply:
Sorry if i misunderstood how you wanted this changed... but from what I am trying to do.. I would like to check if a wap snapshot was published with O(1) while keeping the Util class stateless.. This to me requires the state to be maintained in TableMetadata but ensure the public helper method is in `WapUtil` .. do you agree?